api/src/opentrons/protocol_runner/legacy_command_mapper.py

```diff
@@ -191,78 +228,231 @@ def _build_initial_command(
+        location = command["payload"]["location"]
+        if isinstance(location, LegacyWell):
+            well = location
+        elif location.labware.is_well:
+            well = location.labware.as_well()
+        mount = MountType(pipette.mount)
+        #   the following type checking suppression assumes the tiprack is not loaded on top of a module
+        slot = DeckSlotName.from_primitive(well.parent.parent)  # type: ignore[arg-type]
+        well_name = well.well_name
+        labware_id = self._labware_id_by_slot[slot]
+        pipette_id = self._pipette_id_by_mount[mount]
+        return pe_commands.DropTip.construct(
+            id=command_id,
+            key=command_id,
+            status=pe_commands.CommandStatus.RUNNING,
+            createdAt=now,
+            startedAt=now,
+            params=pe_commands.DropTipParams.construct(
+                pipetteId=pipette_id,
+                labwareId=labware_id,
+                wellName=well_name,
+            ),
+        )
+
+    def _build_pick_up_tip_command(
+        self,
+        command: legacy_command_types.PickUpTipMessage,
+        command_id: str,
+        now: datetime,
+    ) -> pe_commands.Command:
+        pipette: LegacyPipetteContext = command["payload"]["instrument"]
+        location = command["payload"]["location"]
+        if isinstance(location, LegacyWell):
+            well = location
+        elif location.labware.is_well:
+            well = location.labware.as_well()
+        mount = MountType(pipette.mount)
+        #   the following type checking suppression assumes the tiprack is not loaded on top of a module
+        slot = DeckSlotName.from_primitive(well.parent.parent)  # type: ignore[arg-type]
+        well_name = well.well_name
+        labware_id = self._labware_id_by_slot[slot]
+        pipette_id = self._pipette_id_by_mount[mount]
+
+        return pe_commands.PickUpTip.construct(
+            id=command_id,
+            key=command_id,
+            status=pe_commands.CommandStatus.RUNNING,
+            createdAt=now,
+            startedAt=now,
+            params=pe_commands.PickUpTipParams.construct(
+                pipetteId=pipette_id,
+                labwareId=labware_id,
+                wellName=well_name,
+            ),
+        )
+
+    def _build_liquid_handling_commands(
+        self,
+        command: Union[
+            legacy_command_types.AspirateMessage, legacy_command_types.DispenseMessage
+        ],
+        command_id: str,
+        now: datetime,
+    ) -> pe_commands.Command:
+        pipette: LegacyPipetteContext = command["payload"]["instrument"]
+        location = command["payload"]["location"]
+        volume = command["payload"]["volume"]
+        flow_rate = command["payload"]["rate"]
+        if isinstance(location, Location) and location.labware.is_empty is False:
+            if location.labware.is_well:
                 well = location.labware.as_well()
```

Comment:
Let's combine these conditionals.

```suggestion
        if isinstance(location, Location) and location.labware.is_well:
            well = location.labware.as_well()
```

This is important because this line below *requires* `well` to be defined, and this inner conditional is the only place that defines it.

```python
well_name = well.well_name
```

So there's a hazard if `well_name = well.well_name` can ever run without `well = location.labware.as_well()` having run.